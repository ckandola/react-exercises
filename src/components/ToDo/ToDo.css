.finished {
  text-decoration: line-through;
}

.buttons {
  text-align: right;
  padding-right: 2em;
}
