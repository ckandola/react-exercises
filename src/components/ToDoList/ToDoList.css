.todo-background {
  background-color:seashell;
  background-image: url(../../assets/notebook_paper.png);
  max-width: 75vw;
  margin:auto;
  margin-top: 3em;
  align-items: center;
}

.todo-header {
  text-align: center;
  text-decoration: underline;
}

.todo-textEntry {
  text-align: center;
}

.todo-entry {
  width: 20em;
  max-height: 2em;
  padding:0em;
  padding-left: 0.5em;
  background: silver;
}
