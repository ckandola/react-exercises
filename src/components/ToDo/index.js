import React, { useState } from "react";
import "./ToDo.css";

const ToDo = ({ task, onDelete, onEdit }) => {
  const [isFinished, setIsFinished] = useState(false);
  const [isEditing, setIsEditing] = useState(false);
  const [text, setText] = useState(task);


  const onEditTask = () => {
    if (isEditing) {
      onEdit(text, task);
    }
    setIsEditing(!isEditing);
  };

  const onChange = event => {
    setText(event.target.value);
  };

  return (
    <div style={{ display: "flex", textAlign: "left", margin: "0.5em" }}>
      <div className="todo-buttons">
        <input type="submit" value={isEditing ? "Save Changes" : "Edit"} onClick={onEditTask} className="todo-input"/>
      </div>
      {!isEditing ? (
        <>
          <div className="todo-buttons">
            <input
              type="submit"
              value="Delete"
              className="todo-input"
              onClick={() => onDelete(task)}
            />
          </div>
          <input className="todo-check" type="checkbox" value={task} onClick={() => setIsFinished(!isFinished)} />
          {!isFinished ? (
            <span style={{marginLeft: "-1em"}}>{task}</span>
          ) : (
            <span className="todo-finished">{task}</span>
          )}
        </>
      ) : (
        <input style={{background: 'silver'}} value={text} onChange={onChange} />
      )}
    </div>
  );
};

export default ToDo;
