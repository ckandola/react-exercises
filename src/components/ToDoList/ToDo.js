import React, { useState } from "react";
import "./ToDo.css";

const ToDo = ({ task, onDelete, onEdit }) => {
  const [isFinished, setIsFinished] = useState(false);
  const [isEditing, setIsEditing] = useState(false);
  const [text, setText] = useState(task);


  const onEditTask = () => {
    if (isEditing) {
      onEdit(text, task);
    }
    setIsEditing(!isEditing);
  };

  const onChange = event => {
    setText(event.target.value);
  };

  return (
    <div className="todo-main">
      <div className="todo-buttons">
        <input type="submit" value={isEditing ? "Save Changes" : "Edit"} onClick={onEditTask} className="todo-submit"/>
      </div>
      {!isEditing ? (
        <>
          <div className="todo-buttons">
            <input
              type="submit"
              value="Delete"
              className="todo-submit"
              onClick={() => onDelete(task)}
            />
          </div>
          <input className="todo-check" type="checkbox" value={task} onClick={() => setIsFinished(!isFinished)} />
          {!isFinished ? (
            <span>{task}</span>
          ) : (
            <span className="todo-finished">{task}</span>
          )}
        </>
      ) : (
        <input className="todo-entry" value={text} onChange={onChange} />
      )}
    </div>
  );
};

export default ToDo;
