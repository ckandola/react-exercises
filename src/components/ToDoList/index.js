import React, { useState } from "react";
import ToDo from "../ToDo";
import "./ToDoList.css";

const ToDoList = () => {
  const [text, setText] = useState("");
  const [todos, setTodos] = useState([
    "study",
    "work",
    "clean",
    "dust",
    "feed the pets",
    "eat"
  ]);
  let textInput = React.createRef();

  const onSubmit = () => {
    if (text && !todos.includes(text)) {
      const todoCopy = todos.slice();
      todoCopy.push(text);
      setTodos(todoCopy);
      setText("");
      textInput.current.focus();
    }
  };

  const onDelete = taskName => {
    const todoCopy = todos.filter(x => x !== taskName);
    setTodos(todoCopy);
  };

  const onChange = event => {
    setText(event.target.value);
  };

  const onEdit = (newTask, oldTask) => {
    const todoCopy = todos.slice();
    const index = todoCopy.indexOf(oldTask);
    todoCopy[index] = newTask;
    setTodos(todoCopy);
  };

  return (
    <div>
      <h1 className="header">React Header:</h1>
      <ul>
        {todos.map(item => {
          return (
            <ToDo key={item} task={item} onDelete={onDelete} onEdit={onEdit} />
          );
        })}
      </ul>
      <div className="textEntry">
        <input
          className="entry"
          type="text"
          onChange={onChange}
          value={text}
          ref={textInput}
        />
        <input type="submit" value="Add New Task" onClick={onSubmit} />
      </div>
    </div>
  );
};

export default ToDoList;
