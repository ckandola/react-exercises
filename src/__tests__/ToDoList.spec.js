import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { configure, shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import ToDoList from '../components/ToDoList';
import ToDo from '../components/ToDoList/ToDo';

configure({adapter: new Adapter()});

describe('ToDoList tests', () => {
    const wrapper = shallow(<ToDoList />);
    it('ToDoList matches the snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('calls onDelete when Delete is pressed', () => {
        const onDeleteSpy = jest.fn();
        const todoWrapper = shallow(<ToDo task="test a mock function" onDelete={onDeleteSpy} onEdit={() => {}} />);
        const deleteBtn = todoWrapper.find('.todo-submit[value="Delete"]');
        deleteBtn.simulate('click');
        expect(onDeleteSpy).toHaveBeenCalled();
    });

    it('calls onEdit when Edit is pressed and Save Changes is pressed', () => {
        const onEditSpy = jest.fn();
        const todoWrapper = shallow(<ToDo task="Edit something" onDelete={() => {}} onEdit={onEditSpy} />);
        todoWrapper.find('[value="Edit"]').simulate('click');
        todoWrapper.find('.todo-entry').simulate('change', {target: {value: "Edit everything"}});
        todoWrapper.find('[value="Save Changes"]').simulate('click');
        expect(onEditSpy.mock.calls).toHaveLength(1);
    });

    it('adds a new ToDo item when Add New task is clicked', () => {
        const mWrapper = mount(<ToDoList />);
        mWrapper.find('.todo-entry').simulate('change', {target: {value: 'New Task'}});
        mWrapper.find('input[value="Add New Task"]').simulate('click');
        console.log(mWrapper.find('ToDo').getElements())
        expect(mWrapper.find('ToDo')).toHaveLength(7);
    });

    it('crosses off tasks when checkbox is clicked', () => {

    });
});
