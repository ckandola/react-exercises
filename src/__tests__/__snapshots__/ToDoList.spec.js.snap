// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList tests ToDoList matches the snapshot 1`] = `
<div
  className="todo-background"
>
  <h2
    className="todo-header"
  >
    ToDo List:
  </h2>
  <ul>
    <ToDo
      key="study"
      onDelete={[Function]}
      onEdit={[Function]}
      task="study"
    />
    <ToDo
      key="work"
      onDelete={[Function]}
      onEdit={[Function]}
      task="work"
    />
    <ToDo
      key="clean"
      onDelete={[Function]}
      onEdit={[Function]}
      task="clean"
    />
    <ToDo
      key="dust"
      onDelete={[Function]}
      onEdit={[Function]}
      task="dust"
    />
    <ToDo
      key="feed the pets"
      onDelete={[Function]}
      onEdit={[Function]}
      task="feed the pets"
    />
    <ToDo
      key="eat"
      onDelete={[Function]}
      onEdit={[Function]}
      task="eat"
    />
  </ul>
  <div
    className="todo-textEntry"
  >
    <input
      className="todo-entry"
      onChange={[Function]}
      type="text"
      value=""
    />
    <input
      onClick={[Function]}
      type="submit"
      value="Add New Task"
    />
  </div>
</div>
`;
